### PR TITLE
Updated encryption flow in send message and added contact manager error

### PR DESCRIPTION
**Pull request by Laursen79**

### CODE REVIEW

File: client/src/client.rs [hunk 1]
```diff
@@ -171,11 +182,180 @@ impl Client {
...
+                    .body(message.to_owned())
+                    .contact(vec![])
+                    .body_ranges(vec![])
+                    .preview(vec![])
+                    .attachments(vec![])
+                    .build(),
+            )
+            .build();
+
+        // pad and encrypt message.
```
Comment (MJHC) [inline client/src/client.rs:199]: where?

File: client/src/client.rs [hunk 2]
```diff
@@ -171,11 +182,180 @@ impl Client {
...
+                        }
+                        CiphertextMessage::PreKeySignalMessage(_) => {
+                            envelope::Type::PrekeyBundle.into()
+                        }
+                        CiphertextMessage::PlaintextContent(_) => {
+                            envelope::Type::PlaintextContent.into()
+                        }
+                    },
+                    destination_device_id: id.into(),
+                    destination_registration_id: todo!(),
```
Comment (MJHC) [inline client/src/client.rs:271]: todo?

File: client/src/client.rs [hunk 3]
```diff
@@ -171,11 +182,180 @@ impl Client {
...
+                    content: BASE64_STANDARD.encode(msg.serialize()),
+                })
+                .collect(),
+            online: false, // Should this be true?
+            urgent: true,
+            timestamp: timestamp
```
Comment (MJHC) [inline client/src/client.rs:277]: use net_helper func instead if it is in common

File: client/src/client.rs [hunk 4]
```diff
@@ -171,11 +182,180 @@ impl Client {
...
+        let bytes = BASE64_STANDARD
+            .decode(ciphertext_content)
+            .map_err(|err| ReceiveMessageError::Base64DecodeError(err))?;
+
+        // The evelope contains information about which message type is received.
+        let t_id = envelope
+            .r#type
+            .ok_or(ReceiveMessageError::NoMessageTypeInEnvelope)?;
+        let _type = match t_id {
+            2 => Ok(CiphertextMessageType::Whisper),
```
Comment (MJHC) [inline client/src/client.rs:303]: do envelope enum instead of numbers

File: client/src/client.rs [hunk 5]
```diff
@@ -171,11 +182,180 @@ impl Client {
...
+            3 => Ok(CiphertextMessageType::PreKey),
+            7 => Ok(CiphertextMessageType::SenderKey),
+            8 => Ok(CiphertextMessageType::Plaintext),
+            _ => Err(ReceiveMessageError::InvalidMessageTypeInEnvelope(t_id)),
+        }?;
+
+        // Use the information from envelope to construct a CiphertextMessage.
+        let ciphertext = decode_ciphertext(bytes, _type)
```
Comment (MJHC) [inline client/src/client.rs:311]: parse instead of decode maybe?

**Review by MJHC — CHANGES_REQUESTED** [review]
please look at android app and also fix comments

File: client/src/encryption.rs [hunk 6]
```diff
@@ -40,39 +45,85 @@ pub async fn decrypt<R: Rng + CryptoRng>(
...
+
+    if message_length_with_terminator % 160 != 0 {
+        message_part_count += 1;
+    }
+
+    message_part_count * 160
+}
+
+pub fn pad_message(message: &[u8]) -> Vec<u8> {
+    let len = get_padded_message_length(message.len() + 1) - 1;
```
Comment (Mikkel-Boje) [inline client/src/encryption.rs:82]: Why +1 in args, and in the get_padded... func?

File: client/src/encryption.rs [hunk 7]
```diff
@@ -40,39 +45,85 @@ pub async fn decrypt<R: Rng + CryptoRng>(
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Could not unpad message - missing termination char 0x80")
+    }
+}
+
+impl Error for PaddingError {}
+
+fn get_padded_message_length(length: usize) -> usize {
+    let message_length_with_terminator = length + 1;
+    let mut message_part_count = message_length_with_terminator.div_euclid(160);
```
Comment (Mikkel-Boje) [inline client/src/encryption.rs:72]: This is just int division: https://doc.rust-lang.org/core/primitive.usize.html#method.div_euclid

File: client/src/errors.rs [hunk 8]
```diff
@@ -15,17 +23,45 @@ impl fmt::Debug for SignalClientError {
...
 
 impl fmt::Display for SignalClientError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::ContactManagerError(err) => write!(f, "{err}"),
+            Self::RegistrationError(err) => write!(f, "{err}"),
+            Self::LoginError(err) => write!(f, "{err}"),
+            Self::SendMessageError(err) => write!(f, "{err}"),
+            Self::ReceiveMessageError(err) => write!(f, "{err}"),
+        }
```
Comment (Mikkel-Boje) [inline client/src/errors.rs:33]:
```suggestion
        match self {
           _err => write!(f, "{_err}"),
        }
```

File: client/src/errors.rs [hunk 9]
```diff
@@ -119,3 +157,60 @@ impl From<SendMessageError> for SignalClientError {
...
+            Self::CiphertextDecodeError(err) => format!("{err}"),
+            Self::ParseProtocolAddressError(err) => format!("{err}"),
+            Self::DecryptMessageError(err) => format!("{err}"),
+            Self::ProtobufDecodeContentError(err) => {
+                format!("The decrypted message content could not be decoded: {err}")
+            }
+            Self::InvalidMessageContent => {
+                "The message content did not contain a DataMessage".to_owned()
+            }
+        };
```
Comment (Mikkel-Boje) [inline client/src/errors.rs:197]: Maybe do this instead, as such adding more options would also be easy
```
        let message = match self {
            Self::NoMessageTypeInEnvelope => {
                "A message was received but it did not contain the type of the message.".to_owned()
            }
            Self::InvalidMessageTypeInEnvelope(t) => {
                format!("A message was received but it had an invalid message type: {t}")
            }
            Self::ProtobufDecodeContentError(err) => {
                format!("The decrypted message content could not be decoded: {err}")
            }
            Self::InvalidMessageContent => {
                "The message content did not contain a DataMessage".to_owned()
            }
            _err => {
                format!("{_err}"),
            }
        };
```

**Review by Mikkel-Boje — APPROVED** [review]
Look good, but look over my comments and see if you want to implement any of them, if not just resolve :+1:

**Pull request closed without merging**